internal/amos/amos_test.go

```diff
@@ -0,0 +1,478 @@
+package amos
+
+import (
+	"math"
+	"math/rand"
+	"strconv"
+	"testing"
+
+	"github.com/gonum/mathext/internal/amos/amoslib"
+)
+
+type input struct {
+	x    []float64
+	is   []int
+	kode int
+	id   int
+	yr   []float64
+	yi   []float64
+	n    int
+	tol  float64
+}
+
+func randnum(rnd *rand.Rand) float64 {
+	r := 2e2 // Fortran has infinite loop if this is set higher than 2e3
+	if rnd.Float64() > 0.99 {
+		return 0
+	}
+	return rnd.Float64()*r - r/2
+}
+
+func randInput(rnd *rand.Rand) input {
+	x := make([]float64, 8)
+	for j := range x {
+		x[j] = randnum(rnd)
+	}
+	is := make([]int, 3)
+	for j := range is {
+		is[j] = rand.Intn(1000)
+	}
+	kode := rand.Intn(2) + 1
+	id := rand.Intn(2)
+	n := rand.Intn(5) + 1
+	yr := make([]float64, n+1)
+	yi := make([]float64, n+1)
+	for j := range yr {
+		yr[j] = randnum(rnd)
+		yi[j] = randnum(rnd)
+	}
+	tol := 1e-14
+
+	return input{
+		x, is, kode, id, yr, yi, n, tol,
+	}
+}
+
+const nInputs = 100000
+
+func TestAiry(t *testing.T) {
+	rnd := rand.New(rand.NewSource(1))
+	for i := 0; i < nInputs; i++ {
+		in := randInput(rnd)
+		zairytest(t, in.x, in.kode, in.id)
+	}
+}
+
+func TestZacai(t *testing.T) {
+	rnd := rand.New(rand.NewSource(1))
+	for i := 0; i < nInputs; i++ {
+		in := randInput(rnd)
+		zacaitest(t, in.x, in.is, in.tol, in.n, in.yr, in.yi, in.kode)
+	}
+}
+
+func TestZbknu(t *testing.T) {
+	rnd := rand.New(rand.NewSource(1))
+	for i := 0; i < nInputs; i++ {
+		in := randInput(rnd)
+		zbknutest(t, in.x, in.is, in.tol, in.n, in.yr, in.yi, in.kode)
+	}
+}
+
+func TestZasyi(t *testing.T) {
+	rnd := rand.New(rand.NewSource(1))
+	for i := 0; i < nInputs; i++ {
+		in := randInput(rnd)
+		zasyitest(t, in.x, in.is, in.tol, in.n, in.yr, in.yi, in.kode)
+	}
+}
+
+func TestZseri(t *testing.T) {
+	rnd := rand.New(rand.NewSource(1))
+	for i := 0; i < nInputs; i++ {
+		in := randInput(rnd)
+		zseritest(t, in.x, in.is, in.tol, in.n, in.yr, in.yi, in.kode)
+	}
+}
+
+func TestZmlri(t *testing.T) {
+	rnd := rand.New(rand.NewSource(1))
+	for i := 0; i < nInputs; i++ {
+		in := randInput(rnd)
+		zmlritest(t, in.x, in.is, in.tol, in.n, in.yr, in.yi, in.kode)
+	}
+}
+
+func TestZkscl(t *testing.T) {
+	rnd := rand.New(rand.NewSource(1))
+	for i := 0; i < nInputs; i++ {
+		in := randInput(rnd)
+		zkscltest(t, in.x, in.is, in.tol, in.n, in.yr, in.yi)
+	}
+}
+
+func TestZuchk(t *testing.T) {
+	rnd := rand.New(rand.NewSource(1))
+	for i := 0; i < nInputs; i++ {
+		in := randInput(rnd)
+		zuchktest(t, in.x, in.is, in.tol)
+	}
+}
+
+func TestZs1s2(t *testing.T) {
+	rnd := rand.New(rand.NewSource(1))
+	for i := 0; i < nInputs; i++ {
+		in := randInput(rnd)
+		zs1s2test(t, in.x, in.is)
+	}
+}
+
+func zs1s2test(t *testing.T, x []float64, is []int) {
+	ZRR := x[0]
+	ZRI := x[1]
+	S1R := x[2]
+	S1I := x[3]
+	S2R := x[4]
+	S2I := x[5]
+	ASCLE := x[6]
+	ALIM := x[7]
+
+	NZ := is[0]
+	IUF := is[1]
+
+	ZRRfort, ZRIfort, S1Rfort, S1Ifort, S2Rfort, S2Ifort, NZfort, ASCLEfort, ALIMfort, IUFfort :=
```

Comment:
I think in these comparison calls it would be nice to have something that makes it completely clear that the calls are essentillay identical. This means doing the call in a loop and defining a struct type to hold the func and capture the returns.

```
func zs1s2test(t *testing.T, x []float64, is []int) {
	ZRR := x[0]
	ZRI := x[1]
	S1R := x[2]
	S1I := x[3]
	S2R := x[4]
	S2I := x[5]
	ASCLE := x[6]
	ALIM := x[7]

	NZ := is[0]
	IUF := is[1]

	const (
		fort = iota
		amos
	)
	type test struct {
		fn func(
			ZRR, ZRI, S1R, S1I, S2R, S2I float64,
			NZ int, ASCLE, ALIM float64, IUF int,
		) (
			ZRRout, ZRIout, S1Rout, S1Iout, S2Rout, S2Iout float64,
			NZout int,
			ASCLEout, ALIMout float64,
			IUFout int,
		)

		ZRR, ZRI, S1R, S1I, S2R, S2I, ASCLE, ALIM float64
		NZ, IUF                                   int
	}

	results := []test{
		{fn: amoslib.Zs1s2Fort},
		{fn: Zs1s2},
	}
	for i, call := range results {
		results[i].ZRR, results[i].ZRI, results[i].S1R, results[i].S1I, results[i].S2R, results[i].S2I, results[i].NZ, results[i].ASCLE, results[i].ALIM, results[i].IUF =
			call.fn(ZRR, ZRI, S1R, S1I, S2R, S2I, NZ, ASCLE, ALIM, IUF)
	}

	sameF64(t, "zs1s2 zrr", results[fort].ZRR, results[amos].ZRR)
	sameF64(t, "zs1s2 zri", results[fort].ZRI, results[amos].ZRI)
	sameF64(t, "zs1s2 s1r", results[fort].S1R, results[amos].S1R)
	sameF64(t, "zs1s2 s1i", results[fort].S1I, results[amos].S1I)
	sameF64(t, "zs1s2 s2r", results[fort].S2R, results[amos].S2R)
	sameF64(t, "zs1s2 s2i", results[fort].S2I, results[amos].S2I)
	sameF64(t, "zs1s2 ascle", results[fort].ASCLE, results[amos].ASCLE)
	sameF64(t, "zs1s2 alim", results[fort].ALIM, results[amos].ALIM)
	sameInt(t, "iuf", results[fort].IUF, results[amos].IUF)
	sameInt(t, "nz", results[fort].NZ, results[amos].NZ)
}
```

If the result are kept in an embedded struct, then you can use direct value equality tests, but then the output is less informative for failure.